### PR TITLE
Add nullable annotations to StringOps

### PR DESCRIPTION
**Pull request by BCSharp**
Sorry for a large commit; things turned out to be more interdependent than I expected. The `str` API is still not quite done, even with null handling; I have fixed only the obvious issues.

----

It turns out that my changes in `PythonCalls` were not exactly the same as in #789: I didn't have `[NotNull]` on `func` objects. Why is it needed?

### CODE REVIEW

**Comment by slozier** [comment]
> I didn't have [NotNull] on func objects. Why is it needed?

Hmm, possibly not strictly needed? I might have thrown it on there just in case it would be useful...

**Comment by slozier** [comment]
Hmm, someone odd with the PR, it's reverting all the builtin stuff?

**Comment by BCSharp** [comment]
Strange, it builds just fine locally. I will check on it later.

File: Src/IronPython/Runtime/Operations/StringOps.cs [hunk 1]
```diff
@@ -286,7 +290,7 @@ public static class StringOps {
         }
 
         [StaticExtensionMethod]
-        public static object __new__(CodeContext/*!*/ context, PythonType cls, [BytesLike, NotNull]IList<byte> @object, [NotNull]string encoding, [NotNull, DisallowNull]string errors = null) {
+        public static object __new__(CodeContext/*!*/ context, [NotNull]PythonType cls, [BytesLike, NotNull]IList<byte> @object, [NotNull]string encoding, [NotNull, DisallowNull]string? errors = null) {
             if (cls == TypeCache.String) {
                 if (@object is Bytes) return ((Bytes)@object).decode(context, encoding, errors);
```
[the remark this answers is not in the document]
Comment (BCSharp) [reply Src/IronPython/Runtime/Operations/StringOps.cs:295]: I used `null` as default here because it avoids the switch on string, but I have an idea to postpone the whole switch until a decoding error occurs. Then this little efficiency will not matter, overshadowed by the exception handling costs. When done, I will probably replace `null` with `"strict"` as default everywhere in `_codecs` as it is more readable.

File: Src/IronPython/Runtime/Operations/StringOps.cs [hunk 2]
```diff
@@ -23,6 +25,10 @@
 
 using SpecialNameAttribute = System.Runtime.CompilerServices.SpecialNameAttribute;
 using DisallowNullAttribute = System.Diagnostics.CodeAnalysis.DisallowNullAttribute;
+using NotNullWhenAttribute = System.Diagnostics.CodeAnalysis.NotNullWhenAttribute;
```
[the remark this answers is not in the document]
Comment (BCSharp) [reply Src/IronPython/Runtime/Operations/StringOps.cs:28]: Will see how it works.

`IronPython.Hosting`? Where did it come from? I didn't add it and see no reason why it might be needed. I did use the VS refactoring function a couple of times (symbol rename), it must have added it then automatically, don't know why.

File: Src/IronPython/Runtime/Operations/PythonOps.cs [hunk 3]
```diff
@@ -235,7 +241,10 @@ public static partial class PythonOps {
 
             PerfTrack.NoteEvent(PerfTrack.Categories.Temporary, "Repr " + o.GetType().FullName);
 
-            return PythonContext.InvokeUnaryOperator(context, UnaryOperators.Repr, o) as string;
+            object? repr = PythonContext.InvokeUnaryOperator(context, UnaryOperators.Repr, o);
+            if (repr is string strRepr) return strRepr;
+
```
[the remark this answers is not in the document]
Comment (BCSharp) [reply Src/IronPython/Runtime/Operations/PythonOps.cs:246]: Should it **always** return an extensible string, or only if `__repr__` returned one?

[the remark this answers is not in the document]
Comment (slozier) [reply Src/IronPython/Runtime/Operations/PythonOps.cs:246]: No, it should return a normal string pretty much all the time, but in CPython you can write something like:
```Python
class test(str):
    def __repr__(self):
        return self

assert type(repr(test())) == test
```

Changing the method signature to return an object would be quite inconvenient which is why I proposed "unboxing" the extensible string. This might be one of those edge cases which is not worth supporting...

[the remark this answers is not in the document]
Comment (BCSharp) [reply Src/IronPython/Runtime/Operations/PythonOps.cs:246]: I see. And indeed, almost all callers would like to receive a normal string. This can be done with:
```C#
public static T Repr<T>(CodeContext context, object? o) {
    if (o == null) return (T)(object)"None";
    ... // etc.
}

public static string Repr(CodeContext context, object? o)
    => Repr<string>(context, o);
```
and in builtins:
```C#
public static object repr(CodeContext context, object? o)
    => PythonOps.Repr<object>(context, o);
```
However, it will not help the example you posted because `class test(str)` is being fast-tracked as `ICodeFormatable f` into calling `f.__repr__(context)`, which always returns a normal string. I am ready to delete the fast track code (which would fix the your example), except for the warning comment atop:
> could be a container object, we need to detect recursion, but only
> for our own built-in types that we're aware of.  The user can setup
> infinite recursion in their own class if they want.

A few questions:
1. What "our own built-in types that we're aware of" could produce infinite recursion if this fast track is removed? I run all enabled tests but they did not detect any problems. OTOH, the coverage is not very high.
2. It seems that the fast track is catching also user-defined types, contrary to what the comment says? And even then, it does not help against infinite recursion, at least not in those cases I tested.

As for infinite recursion in user-defined types: CPython also does not support it. The run ends with `RuntimeError: maximum recursion depth exceeded while calling a Python object`

[the remark this answers is not in the document]
Comment (BCSharp) [reply Src/IronPython/Runtime/Operations/PythonOps.cs:246]: Another variant:
```C#
internal static object GetReprObject(CodeContext/*!*/ context, object? o) {
    // content of old PythonOps.Repr()
}

public static string Repr(CodeContext/*!*/ context, object? o)
    => (string)GetReprObject(context, o);
```
And in builtins:
```C#
public static object repr(CodeContext/*!*/ context, object? o)
    => PythonOps.GetReprObject(context, o);
```

[the remark this answers is not in the document]
Comment (slozier) [reply Src/IronPython/Runtime/Operations/PythonOps.cs:248]: Personally I'm fine with always returning a string, just because the example I provided is valid, it doesn't mean that it's on purpose? Unless we can find a test or discussion to that effect...

I guess the example I should have given is:
```Python
class test(str): pass

class test2:
  def __repr__(self):
    return test("A")

assert repr(test2()) == "A"
```

If we can support that scenario (I think my code suggestion would cover it), then I think it's enough and we can keep the nicer string returning signature.

---

If we remove the fast track then `ICodeFormattable` seems to serve no purpose beyond providing a method signature? As an alternative to removing the fast track, we could just remove the interface from `ExtensibleString`. But as I mention above it seems like an edge case that is probably not worth supporting.

1. Only example of infinite recursion in repr I can think of would be a collection containing itself. For example `l = list(); l.append(l)`. But I don't see how having a fast track helps with that?
2. I guess it's catching user defined types if they derive from base types that implement `ICodeFormattable`.

Looking at some [discussion](https://mail.python.org/pipermail/ironpython-users/2008-May/007189.html) it seems like `ICodeFormattable` was the only way to get a repr of a .NET object so maybe it's just a leftover mechanism to support that?

By default .NET will throw a `StackOverflowException`, but there's a mechanism to avoid this when running with the command line argument `-X:MaxRecursion`. If you want an example of how this is done you can look at `PythonList.__repr__`.

[the remark this answers is not in the document]
Comment (BCSharp) [reply Src/IronPython/Runtime/Operations/PythonOps.cs:248]: Correction:
```C#
public static string Repr(CodeContext context, object? o)
    => GetReprObject(context, o) switch {
        string str => str,
        Extensible<string> es => es,
        _ => throw new NotImplementedException(),
    };
```

[the remark this answers is not in the document]
Comment (slozier) [reply Src/IronPython/Runtime/Operations/PythonOps.cs:248]: The `GetReprObject` variant is somewhat more potable than the generic version.

[the remark this answers is not in the document]
Comment (slozier) [reply Src/IronPython/Runtime/Operations/PythonOps.cs:248]: Or just?
```Python
public static string Repr(CodeContext context, object? o)
    => GetReprObject(context, o).ToString();
```

[the remark this answers is not in the document]
Comment (BCSharp) [reply Src/IronPython/Runtime/Operations/PythonOps.cs:248]: I know you are fine with just returning a string, but for me it was just too tempting trying to solve the puzzle.  That was fun.  Anyway, the latest commit has the changes you suggested, so it could go onto master, I'm going to set the PR into "ready for merge" state.

I'm OK to leave it like this for now, until a test from StdLib reopens it, or somebody files an issue.  In that case, we know we have a working solution available with the `ICodeFormattable` fast track removed: I tested it with your latest suggestion (just add `!` after `ToString()`) and it seems to work in all cases. Even `l = list(); l.append(l); print(repr(l))` produces `[[...]]` (with or without the fast track), which is the same as in CPython.  Judging from the discussion you referred to, maybe some folks would be happy with `ICodeFormattable` gone.

**Comment by slozier** [comment]
Thanks!
